src/server/frontend_wayland/window_wl_surface_role.cpp

```diff
@@ -410,13 +410,15 @@ void mf::WindowWlSurfaceRole::surface_destroyed()
 {
     if (!client->is_being_destroyed())
     {
+        mark_destroyed();
+
         // "When a client wants to destroy a wl_surface, they must destroy this 'role object' before the wl_surface"
         // NOTE: the wl_shell_surface specification seems contradictory, so this method is overridden in its implementation
         // NOTE: it's also overridden in layer shell, for reasons explained there
-        BOOST_THROW_EXCEPTION(std::runtime_error{
-            "wl_surface@" +
-            (surface ? std::to_string(wl_resource_get_id(surface.value().resource)) : "?") +
-            " destroyed before associated role"});
+        log_warning("wl_surface@%s destroyed before associated role",
+                    (surface ? std::to_string(wl_resource_get_id(surface.value().resource)) : "?").c_str());
+
+        delete this;
```

Comment:
I'm not sure deleting the role is the best way to handle this situation, perhaps we should just leave it until the client does get around to cleaning it up (or until the client closes if it never does)

Reply:
There may be a better way to clean up, but I don't think we should be leaking the object.

A client that subsequently tries to delete, or use, the role would have problems with Mutter (which, as mentioned above, has deleted the id)

Reply:
Specifically, we should probably delete via `wl_resource_destroy(resource)` - except getting to `resource` isn't trivial: typically it is in an inaccessible base class of the derived class.

Reply:
Oh!

So Squeekboard with layer-surface **does** delete it later

Whereas SDL2 with xdg-shell doesn't delete it at all (and couldn't "later" as Mutter has deleted it anyway)

Messy!